src/crystal/system/win32/file_descriptor.cr

```diff
@@ -276,6 +277,59 @@ module Crystal::System::FileDescriptor
   end
 end
 
+private module ConsoleUtils
+  # N UTF-16 code units correspond to no more than 3*N UTF-8 code units.
+  private BUFFER_SIZE = 64
+  @@utf16_buffer = Slice(UInt16).new(BUFFER_SIZE)
+  @@utf8_buffer = Slice(UInt8).new(3 * BUFFER_SIZE)
+
+  # `@@buffer` points to part of `@@utf8_buffer`.
+  # It represents data that has not been read yet.
+  @@buffer : Bytes = @@utf8_buffer[0, 0]
+
+  # Determines if *handle* is a console.
+  def self.console?(handle : LibC::HANDLE) : Bool
+    LibC.GetConsoleMode(handle, out _) != 0
+  end
+
+  # Reads to *slice* from the console specified by *handle*,
+  # and return the actual number of bytes read.
+  def self.read(handle : LibC::HANDLE, slice : Bytes) : Int32
+    return 0 if slice.empty?
+    fill_buffer(handle) if @@buffer.empty?
+
+    bytes_read = slice.size < @@buffer.size ? slice.size : @@buffer.size
+    @@buffer[0, bytes_read].copy_to(slice)
+    @@buffer += bytes_read
+    bytes_read
+  end
+
+  private def self.fill_buffer(handle : LibC::HANDLE) : Nil
+    # Reads in two batches to guarantee that the last character is intact.
+    units_read = read_console(handle, @@utf16_buffer[...-1])
+    return if units_read < 1
+    remainder = @@utf16_buffer + units_read
+    if remainder.size == 1 && @@utf16_buffer[units_read - 1] & 0xFC00 == 0xD800
+      units_read += read_console(handle, remainder)
```

Comment:
Is it ever possible that this read returns 0 bytes, leaving an incomplete UTF-16 code unit sequence at the end of `@@utf16_buffer`, and _then_ a second call to `fill_buffer` puts the remaining part of that sequence at the beginning of `@@utf16_buffer`?

If so then some kind of UTF-16 decoding state persistence is necessary, otherwise a single codepoint could be turned into two replacement characters.

If not then I think there should be a comment saying so, or it could even raise when `read_console` doesn't return 1.